chain/consensus/mir/manager.go

```diff
@@ -0,0 +1,355 @@
+		WAL:                 wal,
+		CryptoManager:       cryptoManager,
+		Net:                 netTransport,
+		InitialValidatorSet: initialValidatorSet,
+		ToMir:               make(chan chan []*mirproto.Request),
+	}
+
+	m.StateManager = NewStateManager(initialMembership, &m)
+
+	mpool := pool.NewModule(
+		m.ToMir,
+		&pool.ModuleConfig{
+			Self:   "mempool",
+			Hasher: "hasher",
+		},
+		&pool.ModuleParams{
+			MaxTransactionsInBatch: 10,
+		},
+	)
+
+	smrSystem, err := smr.New(
+		t.NodeID(mirID),
+		h,
+		initialMembership,
+		m.CryptoManager,
+		m.StateManager,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create SMR system: %w", err)
+	}
+	smrSystem = smrSystem.WithModule("mempool", mpool)
+
+	if err := smrSystem.Start(ctx); err != nil {
+		return nil, fmt.Errorf("could not start SMR system: %w", err)
+	}
+
+	cfg := mir.DefaultNodeConfig().WithLogger(logger)
+
+	interceptorOutput := os.Getenv(InterceptorOutputEnv)
+	if interceptorOutput != "" {
+		// TODO: Persist in repo path?
+		m.interceptor, err = eventlog.NewRecorder(
+			t.NodeID(mirID),
+			interceptorOutput,
+			logging.Decorate(logger, "Interceptor: "),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create interceptor: %w", err)
+		}
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, m.interceptor)
+	} else {
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, nil)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Mir node: %w", err)
+	}
+
+	return &m, nil
+}
+
+// Start starts the manager.
+func (m *Manager) Start(ctx context.Context) chan error {
+	log.Infof("Mir manager %s starting", m.MirID)
+
+	errChan := make(chan error, 1)
+
+	go func() {
+		// Run Mir node until it stops.
+		if err := m.MirNode.Run(ctx); err != nil && !errors.Is(err, mir.ErrStopped) {
+			log.Infof("Mir manager %s: Mir node stopped with error: %v", m.MirID, err)
+			errChan <- err
+		}
+
+		// Perform cleanup of Node's modules.
+		m.Stop()
+	}()
+
+	return errChan
+}
+
+// Stop stops the manager and all its components.
+func (m *Manager) Stop() {
+	log.With("miner", m.MirID).Infof("Mir manager shutting down")
+	defer log.With("miner", m.MirID).Info("Mir manager stopped")
+
+	if err := m.WAL.Close(); err != nil {
+		log.Errorf("Could not close write-ahead log: %s", err)
+	}
+	log.Info("WAL closed")
+
+	if m.interceptor != nil {
+		if err := m.interceptor.Stop(); err != nil {
+			log.Errorf("Could not close interceptor: %s", err)
+		}
+		log.Info("Interceptor closed")
+	}
+
+	m.Net.Stop()
+	log.Info("Network transport stopped")
+}
+
+// ReconfigureMirNode reconfigures the Mir node.
+func (m *Manager) ReconfigureMirNode(ctx context.Context, nodes map[t.NodeID]t.NodeAddress) error {
+	log.With("miner", m.MirID).Debug("Reconfiguring a Mir node")
+
+	if len(nodes) == 0 {
+		return fmt.Errorf("empty validator set")
+	}
+
+	go m.Net.Connect(ctx, nodes)
+	go m.Net.CloseOldConnections(nodes)
```

Comment:
> I thought that the moment we start the reconfiguraton we deemed "useless" the connection of the validators leaving the committee

Well, that's not quite the case. If we are a bit faster than the other nodes, we might already know all we need to transition to the next epoch, but other nodes might not. In fact, since new memberships are currently announced way ahead of time, this would not even work if nodes were actually leaving (their connections would be closed prematurely). `CloseOldConnections` is to be called on reception of a stable checkpoint (that's when we garbage-collect), not on the announcement of a new (future) membership.

But as I said, since for how the membership is static, this doesn't matter in practice. But I'd still remove that line, because principally it doesn't belong there (as closing old connection is part of garbage collection that should happen on checkpoints). Properly invoking `CloseOldConnections` during garbage collection is, however, not as trivial as this, and by the time we actually reconfigure, we might have slightly changed way of managing the network connections, so let's worry about it later and just remove the line for now.

Reply:
Gotcha, I comment the `CloseOldConnections` then and add a message to explain that this should be handled in the future.